cypress/plugins/index.ts

```diff
@@ -0,0 +1,242 @@
+  await new Promise(resolve => setTimeout(resolve, ms));
+};
+
+type NodeId = number;
+
+interface Node {
+  id: NodeId;
+  httpListen: number;
+  peerListen: number;
+  workspacePath: string;
+  pid?: number;
+  authToken?: string;
+  peerAddress?: string;
+}
+
+export interface OnboardedNode {
+  authToken: string;
+  peerAddress: string;
+  workspacePath: string;
+}
+
+export default (
+  on: Cypress.PluginEvents,
+  _config: Cypress.PluginConfigOptions
+): void => {
+  on("task", {
+    // Kill any nodes that were started by this plugin.
+    async killAllNodes() {
+      console.log("  ### killAllNodes()");
+
+      nodes.forEach(async node => {
+        try {
+          await exec(`kill -9 ${node.pid}`);
+        } catch (err) {
+          console.log("    ERROR: ", err.message);
+        }
+      });
+      nodes = [];
+      return null;
+    },
+    // Start a new `radicle-proxy` node, the `id` is used as a port number for
+    // both API TCP connections as well as UDP peer connections.
+    //
+    // This task also sets up an empty workspace directory where we can create
+    // new project and checkout directories or store any other data necessary
+    // for a testcase:
+    //
+    //   ./radicle-upstream/cypress/workspace/<RANDOM_UUID>
+    //
+    // This command waits on the proxy process to start listening on its HTTP
+    // API port before returning.
+    async startNode(id: NodeId) {
+      console.log(`  ### startNode(${id})`);
+
+      const node: Node = {
+        id: id,
+        httpListen: id,
+        peerListen: id,
+        workspacePath: path.join(CYPRESS_WORKSPACE_PATH, uuid.v4()),
+      };
+
+      await fs.mkdirs(node.workspacePath);
+
+      const process = childProcess.spawn(
+        PROXY_PATH,
+        [
+          "--test",
+          "--http-listen",
+          `${HOST}:${node.httpListen}`,
+          "--peer-listen",
+          `${HOST}:${node.peerListen}`,
+        ],
+        {}
+      );
+
+      node.pid = process.pid;
+
+      process.on("exit", async () => {
+        console.log(`    [${id}] node terminated`);
+        if (node.workspacePath) {
+          console.log(
+            `    [${id}] cleaning up temp directory ${node.workspacePath}`
+          );
+          await exec(`rm -rf ${node.workspacePath}`);
+        }
+
+        return null;
+      });
+
+      process.stderr.setEncoding("utf8");
+      process.stderr.on("data", data => {
+        console.log(`    [${id}] STDERR: ${data.trim()}`);
+      });
+
+      process.stdout.setEncoding("utf8");
+      process.stdout.on("data", data => {
+        console.log(`    [${id}] STDOUT: ${data.trim()}`);
+      });
+
+      await waitOn({ resources: [`tcp:${HOST}:${id}`] });
+      console.log(`    [${id}] node started successfully`);
+
+      nodes.push(node);
+
+      return null;
+    },
+    // Onboard a node by requesting and storing an auth token and creating
+    // an identity.
+    async onboardNode(id: NodeId) {
```

Comment:
It does not seem necessary to me to run this and other tasks (except `startNode` and `killAllNodes`) on the backend side of cypress. It should be possible to just turn these into regular functions and not use the `task` mechanism at all.

Reply:
See https://github.com/radicle-dev/radicle-upstream/pull/1492#discussion_r565382281.